example/contrib/socks-client/socks/impl/handshake.hpp

```diff
@@ -0,0 +1,598 @@
+            if (ec)
+                break;
+
+            write<uint32_t>(address.to_uint(), req); // DST I
+
+            if (!username_.empty())
+            {
+                std::copy(username_.begin(), username_.end(), req);    // USERID
+                req += username_.size();
+            }
+            write<uint8_t>(0, req); // NULL.
+
+            request.commit(bytes_to_write);
+            return net::async_write(stream_, request, std::move(*this));
+        }
+        case 1:
+        {
+            step_ = 2;
+            return net::async_read(stream_, response,
+                net::transfer_exactly(8), std::move(*this));
+        }
+        case 2:
+        {
+            auto resp = static_cast<const unsigned char*>(response.data().data());
+
+            read<uint8_t>(resp); // VN is the version of the reply code and should be 0.
+            auto cd = read<uint8_t>(resp);
+
+            if (cd != detail::SOCKS4_REQUEST_GRANTED)
+            {
+                switch (cd)
+                {
+                case detail::SOCKS4_REQUEST_REJECTED_OR_FAILED:
+                    ec = error::socks_request_rejected_or_failed;
+                    break;
+                case detail::SOCKS4_CANNOT_CONNECT_TARGET_SERVER:
+                    ec = error::socks_request_rejected_cannot_connect;
+                    break;
+                case detail::SOCKS4_REQUEST_REJECTED_USER_NO_ALLOW:
+                    ec = error::socks_request_rejected_incorrect_userid;
+                    break;
+                default:
+                    ec = error::socks_unknown_error;
+                    break;
+                }
+            }
+            BOOST_FALLTHROUGH;
+        }
+        case 3:
+            break;
+        }
+
+        this->complete_now(ec);
+    }
+
+private:
+    Stream& stream_;
+
+    using BufferPtr = std::unique_ptr<Buffer>;
+    BufferPtr request_{ new Buffer() }; // std::make_unique c++14 or later.
+    BufferPtr response_{ new Buffer() };
+
+    std::string hostname_;
+    unsigned short port_;
+    std::string username_;
+    int step_ = 0;
+};
+
+template<
+    class Stream,
+    class Handler,
+    class Buffer,
+    class base_type = boost::beast::async_base<
+        Handler, typename Stream::executor_type>>
+class socks5_op : public base_type
+{
+public:
+    socks5_op(socks5_op&&) = default;
+    socks5_op(socks5_op const&) = default;
+
+    socks5_op(
+        Stream& stream,
+        Handler& handler,
+        const std::string& hostname,
+        unsigned short port,
+        const std::string& username,
+        const std::string& password,
+        bool use_hostname)
+        : base_type(std::move(handler), stream.get_executor())
+        , stream_(stream)
+        , hostname_(hostname)
+        , port_(port)
+        , username_(username)
+        , password_(password)
+        , use_hostname_(use_hostname)
+    {
+        (*this)({}, 0); // start the operation
```

Comment:
If we can build the initial request in the constructor that would be best, although it looks like we might need to save the username and password. Although we could do that in a single buffer instead of using two strings.